File: apps/api/AssetManagement.Api/Services/AuditService.cs
using AssetManagement.Api.Data;
using AssetManagement.Api.Models;
using AssetManagement.Api.Models.Enums;

namespace AssetManagement.Api.Services;

public class AuditService(AppDbContext db) : IAuditService
{
    public async Task LogAsync(AuditEntry entry)
    {
        var auditLog = new AuditLog
        {
            Id = Guid.NewGuid(),
            ActorId = entry.ActorId,
            ActorName = entry.ActorName,
            Action = entry.Action,
            EntityType = entry.EntityType,
            EntityId = entry.EntityId,
            EntityName = entry.EntityName,
            Source = AuditSource.API,
            Details = entry.Details,
        };

        db.AuditLogs.Add(auditLog);

        if (entry.EntityType == "Asset" && Guid.TryParse(entry.EntityId, out var assetId))
        {
            var eventType = entry.Action switch
            {
                "Created" => AssetHistoryEventType.Created,
                "Updated" => AssetHistoryEventType.Edited,
                "Archived" => AssetHistoryEventType.Archived,
                "CheckedOut" => AssetHistoryEventType.CheckedOut,
                "CheckedIn" => AssetHistoryEventType.CheckedIn,
                "Retired" => AssetHistoryEventType.Retired,
                "Sold" => AssetHistoryEventType.Sold,
                _ => (AssetHistoryEventType?)null,
            };

            if (eventType is not null)
            {
                var history = new AssetHistory
                {
                    Id = Guid.NewGuid(),
                    AssetId = assetId,
                    EventType = eventType.Value,
                    PerformedByUserId = entry.ActorId,
                    Details = entry.Details,
                };

                if (entry.Changes is { Count: > 0 })
                {
                    foreach (var change in entry.Changes)
                    {
                        history.Changes.Add(new AssetHistoryChange
                        {
                            Id = Guid.NewGuid(),
                            FieldName = change.FieldName,
                            OldValue = change.OldValue,
                            NewValue = change.NewValue,
                        });
                    }
                }

                db.AssetHistory.Add(history);
            }
        }

        if (entry.EntityType == "Certificate" && Guid.TryParse(entry.EntityId, out var certificateId))
        {
            var certEventType = entry.Action switch
            {
                "Created" => CertificateHistoryEventType.Created,
                "Updated" => CertificateHistoryEventType.Edited,
                "Archived" => CertificateHistoryEventType.Archived,
                "Renewed" => CertificateHistoryEventType.Renewed,
                "Revoked" => CertificateHistoryEventType.Revoked,
                _ => (CertificateHistoryEventType?)null,
            };

            if (certEventType is not null)
            {
                var certHistory = new CertificateHistory
                {
                    Id = Guid.NewGuid(),
                    CertificateId = certificateId,
                    EventType = certEventType.Value,
                    PerformedByUserId = entry.ActorId,
                    Details = entry.Details,
                };

                if (entry.Changes is { Count: > 0 })
                {
                    foreach (var change in entry.Changes)
                    {
                        certHistory.Changes.Add(new CertificateHistoryChange
                        {
                            Id = Guid.NewGuid(),
                            FieldName = change.FieldName,
                            OldValue = change.OldValue,
                            NewValue = change.NewValue,
                        });
                    }
                }

                db.CertificateHistory.Add(certHistory);
            }
        }

        if (entry.EntityType == "Application" && Guid.TryParse(entry.EntityId, out var applicationId))
        {
            var appEventType = entry.Action switch
            {
                "Created" => ApplicationHistoryEventType.Created,
                "Updated" => ApplicationHistoryEventType.Edited,
                "Archived" => ApplicationHistoryEventType.Archived,
                "Renewed" => ApplicationHistoryEventType.Renewed,
                "Suspended" => ApplicationHistoryEventType.Suspended,
                _ => (ApplicationHistoryEventType?)null,
            };

            if (appEventType is not null)
            {
                var appHistory = new ApplicationHistory
                {
                    Id = Guid.NewGuid(),
                    ApplicationId = applicationId,
                    EventType = appEventType.Value,
                    PerformedByUserId = entry.ActorId,
                    Details = entry.Details,
                };

                if (entry.Changes is { Count: > 0 })
                {
                    foreach (var change in entry.Changes)
                    {
                        appHistory.Changes.Add(new ApplicationHistoryChange
                        {
                            Id = Guid.NewGuid(),
                            FieldName = change.FieldName,
                            OldValue = change.OldValue,
                            NewValue = change.NewValue,
                        });
                    }
                }

                db.ApplicationHistory.Add(appHistory);
            }
        }

        await db.SaveChangesAsync();
    }
}

File: apps/api-kt/src/main/kotlin/com/assetmanagement/api/service/AuditService.kt
package com.assetmanagement.api.service

import com.assetmanagement.api.model.*
import com.assetmanagement.api.model.enums.*
import com.assetmanagement.api.repository.*
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

data class AuditChange(
    val fieldName: String,
    val oldValue: String?,
    val newValue: String?
)

data class AuditEntry(
    val action: String,
    val entityType: String,
    val entityId: String,
    val entityName: String? = null,
    val details: String? = null,
    val actorId: UUID? = null,
    val actorName: String = "System",
    val changes: List<AuditChange>? = null
)

@Service
class AuditService(
    private val auditLogRepository: AuditLogRepository,
    private val assetHistoryRepository: AssetHistoryRepository,
    private val personHistoryRepository: PersonHistoryRepository,
    private val certificateHistoryRepository: CertificateHistoryRepository,
    private val applicationHistoryRepository: ApplicationHistoryRepository
) {

    @Transactional
    fun log(entry: AuditEntry) {
        val auditLog = AuditLog(
            actorId = entry.actorId,
            actorName = entry.actorName,
            action = entry.action,
            entityType = entry.entityType,
            entityId = entry.entityId,
            entityName = entry.entityName,
            source = AuditSource.API,
            details = entry.details
        )
        auditLogRepository.save(auditLog)

        when (entry.entityType) {
            "Asset" -> logAssetHistory(entry)
            "Person" -> logPersonHistory(entry)
            "Certificate" -> logCertificateHistory(entry)
            "Application" -> logApplicationHistory(entry)
        }
    }

    private fun logAssetHistory(entry: AuditEntry) {
        val assetId = try { UUID.fromString(entry.entityId) } catch (_: Exception) { return }
        val eventType = when (entry.action) {
            "Created" -> AssetHistoryEventType.Created
            "Updated" -> AssetHistoryEventType.Edited
            "Archived" -> AssetHistoryEventType.Archived
            "CheckedOut" -> AssetHistoryEventType.CheckedOut
            "CheckedIn" -> AssetHistoryEventType.CheckedIn
            "Retired" -> AssetHistoryEventType.Retired
            "Sold" -> AssetHistoryEventType.Sold
            else -> return
        }
        val history = AssetHistory(
            assetId = assetId,
            eventType = eventType,
            performedByUserId = entry.actorId,
            details = entry.details
        )
        entry.changes?.forEach { change ->
            val historyChange = AssetHistoryChange(
                assetHistoryId = history.id,
                fieldName = change.fieldName,
                oldValue = change.oldValue,
                newValue = change.newValue
            )
            history.changes.add(historyChange)
        }
        assetHistoryRepository.save(history)
    }

    private fun logPersonHistory(entry: AuditEntry) {
        val personId = try { UUID.fromString(entry.entityId) } catch (_: Exception) { return }
        val eventType = when (entry.action) {
            "Created" -> PersonHistoryEventType.Created
            "Updated" -> PersonHistoryEventType.Edited
            "Archived" -> PersonHistoryEventType.Archived
            "Restored" -> PersonHistoryEventType.Restored
            "AssetAssigned" -> PersonHistoryEventType.AssetAssigned
            "AssetUnassigned" -> PersonHistoryEventType.AssetUnassigned
            "AssetCheckedOut" -> PersonHistoryEventType.AssetCheckedOut
            "AssetCheckedIn" -> PersonHistoryEventType.AssetCheckedIn
            else -> return
        }
        val history = PersonHistory(
            personId = personId,
            eventType = eventType,
            performedByUserId = entry.actorId,
            details = entry.details
        )
        entry.changes?.forEach { change ->
            val historyChange = PersonHistoryChange(
                personHistoryId = history.id,
                fieldName = change.fieldName,
                oldValue = change.oldValue,
                newValue = change.newValue
            )
            history.changes.add(historyChange)
        }
        personHistoryRepository.save(history)
    }

    private fun logCertificateHistory(entry: AuditEntry) {
        val certificateId = try { UUID.fromString(entry.entityId) } catch (_: Exception) { return }
        val eventType = when (entry.action) {
            "Created" -> CertificateHistoryEventType.Created
            "Updated" -> CertificateHistoryEventType.Edited
            "Archived" -> CertificateHistoryEventType.Archived
            "Renewed" -> CertificateHistoryEventType.Renewed
            "Revoked" -> CertificateHistoryEventType.Revoked
            else -> return
        }
        val history = CertificateHistory(
            certificateId = certificateId,
            eventType = eventType,
            performedByUserId = entry.actorId,
            details = entry.details
        )
        entry.changes?.forEach { change ->
            val historyChange = CertificateHistoryChange(
                certificateHistoryId = history.id,
                fieldName = change.fieldName,
                oldValue = change.oldValue,
                newValue = change.newValue
            )
            history.changes.add(historyChange)
        }
        certificateHistoryRepository.save(history)
    }

    private fun logApplicationHistory(entry: AuditEntry) {
        val applicationId = try { UUID.fromString(entry.entityId) } catch (_: Exception) { return }
        val eventType = when (entry.action) {
            "Created" -> ApplicationHistoryEventType.Created
            "Updated" -> ApplicationHistoryEventType.Edited
            "Archived" -> ApplicationHistoryEventType.Archived
            "Renewed" -> ApplicationHistoryEventType.Renewed
            "Suspended" -> ApplicationHistoryEventType.Suspended
            else -> return
        }
        val history = ApplicationHistory(
            applicationId = applicationId,
            eventType = eventType,
            performedByUserId = entry.actorId,
            details = entry.details
        )
        entry.changes?.forEach { change ->
            val historyChange = ApplicationHistoryChange(
                applicationHistoryId = history.id,
                fieldName = change.fieldName,
                oldValue = change.oldValue,
                newValue = change.newValue
            )
            history.changes.add(historyChange)
        }
        applicationHistoryRepository.save(history)
    }
}
